src/backend/xdg_desktop_portal.rs

```diff
@@ -0,0 +1,284 @@
+        for file_extension in &filter.extensions {
+            ashpd_filter = ashpd_filter.glob(&format!("*.{}", file_extension));
+        }
+        options = options.add_filter(ashpd_filter);
+    }
+    options
+}
+
+fn add_filters_to_save_file_options(
+    filters: Vec<Filter>,
+    mut options: SaveFileOptions,
+) -> SaveFileOptions {
+    for filter in &filters {
+        let mut ashpd_filter = FileFilter::new(&filter.name);
+        for file_extension in &filter.extensions {
+            ashpd_filter = ashpd_filter.glob(&format!("*.{}", file_extension));
+        }
+        options = options.add_filter(ashpd_filter);
+    }
+    options
+}
+
+// refer to https://github.com/flatpak/xdg-desktop-portal/issues/213
+fn uri_to_pathbuf(uri: &str) -> Option<PathBuf> {
+    uri.strip_prefix("file://").map(PathBuf::from)
+}
+
+fn unwrap_or_warn<T, E: std::fmt::Debug>(result: Result<T, E>) -> Option<T> {
+    match result {
+        Err(e) => {
+            warn!("{:?}", e);
+            None
+        }
+        Ok(t) => Some(t),
+    }
+}
+
+//
+// File Picker
+//
+
+use crate::backend::FilePickerDialogImpl;
+impl FilePickerDialogImpl for FileDialog {
+    fn pick_file(self) -> Option<PathBuf> {
+        let connection = unwrap_or_warn(block_on(zbus::Connection::session()))?;
+        let proxy = unwrap_or_warn(block_on(FileChooserProxy::new(&connection)))?;
+        let mut options = OpenFileOptions::default()
+            .accept_label("Pick file")
+            .multiple(false);
+        options = add_filters_to_open_file_options(self.filters, options);
+        let selected_files = block_on(proxy.open_file(
+            &WindowIdentifier::default(),
+            &self.title.unwrap_or_else(|| "Pick a file".to_string()),
+            options,
+        ));
+        if selected_files.is_err() {
+            return None;
+        }
+        uri_to_pathbuf(&selected_files.unwrap().uris()[0])
+    }
+
+    fn pick_files(self) -> Option<Vec<PathBuf>> {
+        let connection = unwrap_or_warn(block_on(zbus::Connection::session()))?;
+        let proxy = unwrap_or_warn(block_on(FileChooserProxy::new(&connection)))?;
+        let mut options = OpenFileOptions::default()
+            .accept_label("Pick file")
+            .multiple(true);
+        options = add_filters_to_open_file_options(self.filters, options);
+        let selected_files = block_on(proxy.open_file(
+            &WindowIdentifier::default(),
+            &self.title.unwrap_or_else(|| "Pick a file".to_string()),
+            options,
+        ));
+        if selected_files.is_err() {
+            return None;
+        }
+        let selected_files = selected_files
+            .unwrap()
+            .uris()
+            .iter()
+            .filter_map(|string| uri_to_pathbuf(string))
+            .collect::<Vec<PathBuf>>();
+        if selected_files.is_empty() {
+            return None;
+        }
+        Some(selected_files)
+    }
```

Comment:
Those look like 1 to 1 duplications of async versions, why not just `block_on(async_imp())` so we have one impl for both

Reply:
I can try, but #40 makes that awkward.

Reply:
It's not really a problem, you can just call .path() on async result.

Reply:
Option::map to the rescue: b74ff6f875254780219fc3112f04761a25f08b5b